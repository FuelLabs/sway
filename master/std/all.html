<!DOCTYPE html><html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="swaydoc"><meta name="description" content="List of all items in this project"><meta name="keywords" content="sway, swaylang, sway-lang"><link rel="icon" href="static.files/sway-logo.svg"><title>List of all items in this project</title><link rel="stylesheet" type="text/css" href="static.files/normalize.css"><link rel="stylesheet" type="text/css" href="static.files/swaydoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="static.files/ayu.css"><link rel="stylesheet" href="static.files/ayu.min.css"></head><body class="swaydoc mod"><nav class="sidebar"><a class="sidebar-logo" href="../std/index.html"><div class="logo-container"><img class="sway-logo" src="static.files/sway-logo.svg" alt="logo"></div></a><h2 class="location">Library std</h2><div class="sidebar-elems"><a id="all-types" href="index.html"><p>Back to index</p></a><section><div class="block"><ul><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#constants">Constants</a></li><li><a href="#functions">Functions</a></li><li><a href="#primitives">Primitives</a></li></ul></div></section></div></nav><main><div class="width-limiter"><script src="search.js" type="text/javascript"></script><script>function onSearchFormSubmit(event){event.preventDefault();const searchQuery=document.getElementById("search-input").value;const url=new URL(window.location.href);if(searchQuery){url.searchParams.set('search',searchQuery);}else{url.searchParams.delete('search');}history.pushState({search:searchQuery},"",url);window.dispatchEvent(new HashChangeEvent("hashchange"));}document.addEventListener('DOMContentLoaded',()=>{const searchbar=document.getElementById("search-input");const searchForm=document.getElementById("search-form");searchbar.addEventListener("keyup",function(event){onSearchFormSubmit(event);});searchbar.addEventListener("search",function(event){onSearchFormSubmit(event);});function onQueryParamsChange(){const searchParams=new URLSearchParams(window.location.search);const query=searchParams.get("search");const searchSection=document.getElementById('search');const mainSection=document.getElementById('main-content');const searchInput=document.getElementById('search-input');if(query){searchInput.value=query;const results=Object.values(SEARCH_INDEX).flat().filter(item=>{const lowerQuery=query.toLowerCase();return item.name.toLowerCase().includes(lowerQuery);});const header=`<h1>Results for ${query}</h1>`;if(results.length>0){const resultList=results.map(item=>{const formattedName=`<span class="type ${item.type_name}">${item.name}</span>`;const name=item.type_name==="module"?[...item.module_info.slice(0,-1),formattedName].join("::"):[...item.module_info,formattedName].join("::");const path=["..",...item.module_info,item.html_filename].join("/");const left=`<td><span>${name}</span></td>`;const right=`<td><p>${item.preview}</p></td>`;return`<tr onclick="window.location='${path}';">${left}${right}</tr>`;}).join('');searchSection.innerHTML=`${header}<table>${resultList}</table>`;}else{searchSection.innerHTML=`${header}<p>No results found.</p>`;}searchSection.setAttribute("class","search-results");mainSection.setAttribute("class","content hidden");}else{searchSection.setAttribute("class","search-results hidden");mainSection.setAttribute("class","content");}}window.addEventListener('hashchange',onQueryParamsChange);onQueryParamsChange();});</script><nav class="sub"><form id="search-form" class="search-form" onsubmit="onSearchFormSubmit(event)"><div class="search-container"><input id="search-input" class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Search the docs..." type="search"></div></form></nav><section id="main-content" class="content"><h1 class="fqn"><span class="in-band">List of all items</span></h1><h2 id="structs">Structs</h2><div class="item-table"><div class="item-row"><div class="item-left Struct-item"><a class="struct" href="low_level_call/struct.CallParams.html">low_level_call::CallParams</a></div><div class="item-right docblock-short"><p>A struct representing the call parameters of a function call.</p>
</div></div></div><h2 id="enums">Enums</h2><div class="item-table"><div class="item-row"><div class="item-left Enum-item"><a class="enum" href="auth/enum.AuthError.html">auth::AuthError</a></div><div class="item-right docblock-short"><p>The error type used when an <code>Identity</code> cannot be determined.</p>
</div></div></div><h2 id="traits">Traits</h2><div class="item-table"><div class="item-row"><div class="item-left Trait-item"><a class="trait" href="storage/storable_slice/trait.StorableSlice.html">storage::storable_slice::StorableSlice</a></div><div class="item-right docblock-short"><p>A general way to persistently store heap types.</p>
</div></div></div><h2 id="constants">Constants</h2><div class="item-table"><div class="item-row"><div class="item-left Constant-item"><a class="constant" href="inputs/constant.GTF_INPUT_MESSAGE_PREDICATE_DATA.html">inputs::GTF_INPUT_MESSAGE_PREDICATE_DATA</a></div></div></div><h2 id="functions">Functions</h2><div class="item-table"><div class="item-row"><div class="item-left Function-item"><a class="fn" href="low_level_call/fn.call_with_function_selector.html">low_level_call::call_with_function_selector</a></div><div class="item-right docblock-short"><p>Call a target contract with a function selector and calldata, provided as <code>Bytes</code>.</p></div></div></div><h2 id="primitives">Primitives</h2><div class="item-table"><div class="item-row"><div class="item-left Primitive-item"><a class="primitive" href="primitive.str[64].html">str[64]</a></div><div class="item-right docblock-short">fixed-length string</div></div></div></section><section id="search" class="search-results"></section></div></main><script src="static.files/highlight.js"></script><script>hljs.highlightAll();</script></body></html>