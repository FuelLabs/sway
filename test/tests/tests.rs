use libtest_mimic::{Arguments, Trial};
use normalize_path::NormalizePath;
use regex::Regex;
use std::{path::PathBuf, str::FromStr, sync::Once};

static FORC_COMPILATION: Once = Once::new();
static FORC_DOC_COMPILATION: Once = Once::new();

fn compile_forc() {
    let args = vec!["b", "--release", "-p", "forc"];
    let o = std::process::Command::new("cargo")
        .args(args)
        .output()
        .unwrap();
    assert!(o.status.success());
}

fn compile_forc_doc() {
    let args = vec!["b", "--release", "-p", "forc-doc"];
    let o = std::process::Command::new("cargo")
        .args(args)
        .output()
        .unwrap();
    assert!(o.status.success());
}

pub fn main() {
    let repo_root: PathBuf =
        PathBuf::from_str(&std::env::var("CARGO_MANIFEST_DIR").unwrap()).unwrap();
    let repo_root = repo_root.parent().unwrap().to_path_buf();

    let mut args = Arguments::from_args();
    args.nocapture = true;

    let tests = discover_test()
        .into_iter()
        .map(|dir| {
            let test_programs_dir = "src/e2e_vm_tests/test_programs/";
            let name = dir
                .to_str()
                .unwrap()
                .to_string()
                .replace(test_programs_dir, "");

            let repo_root = repo_root.clone();
            Trial::test(name, move || {
                let snapshot_toml =
                    std::fs::read_to_string(format!("{}/snapshot.toml", dir.display()))?;
                let snapshot_toml = toml::from_str::<toml::Value>(&snapshot_toml)?;
                let cmds = if let Some(cmds) = snapshot_toml.get("cmds") {
                    cmds.as_array()
                        .unwrap()
                        .iter()
                        .map(|v| v.as_str().unwrap())
                        .collect::<Vec<_>>()
                } else {
                    vec!["forc build --path {root}"]
                };

                let root = format!("test/{}", dir.display());

                use std::fmt::Write;
                let mut snapshot = String::new();

                for cmd in cmds {
                    let cmd = cmd.replace("{root}", &root);

                    let _ = writeln!(&mut snapshot, "> {}", cmd);

                    // known commands
                    let cmd = if let Some(cmd) = cmd.strip_prefix("forc doc ") {
                        FORC_DOC_COMPILATION.call_once(|| {
                            compile_forc_doc();
                        });
                        format!("target/release/forc-doc {cmd} 1>&2")
                    } else if let Some(cmd) = cmd.strip_prefix("forc ") {
                        FORC_COMPILATION.call_once(|| {
                            compile_forc();
                        });
                        format!("target/release/forc {cmd} 1>&2")
                    } else {
                        panic!("Not supported. Possible commands: forc")
                    };

                    let o = duct::cmd!("bash", "-c", cmd.clone())
                        .dir(repo_root.clone())
                        .stderr_to_stdout()
                        .stdout_capture()
                        .env("COLUMNS", "10")
                        .unchecked()
                        .start()
                        .unwrap();
                    let o = o.wait().unwrap();

                    let _ = writeln!(
                        &mut snapshot,
                        "{}",
                        clean_output(&format!(
                            "exit status: {}\noutput:\n{}",
                            o.status.code().unwrap(),
                            std::str::from_utf8(&o.stdout).unwrap(),
                        ))
                    );
                }

                fn stdout(root: &str, snapshot: &str) {
                    let root = PathBuf::from_str(root).unwrap();
                    let root = root.normalize();

                    let mut insta = insta::Settings::new();
                    insta.set_snapshot_path(root);
                    insta.set_prepend_module_to_snapshot(false);
                    insta.set_omit_expression(true);
                    let scope = insta.bind_to_scope();
                    insta::assert_snapshot!("stdout", snapshot);
                    drop(scope);
                }
                stdout(&format!("{}/{root}", repo_root.display()), &snapshot);

                Ok(())
            })
        })
        .collect();
    libtest_mimic::run(&args, tests).exit();
}

pub fn discover_test() -> Vec<PathBuf> {
    use glob::glob;

    let mut entries = vec![];

    for entry in glob("**/snapshot.toml")
        .expect("Failed to read glob pattern")
        .flatten()
    {
        entries.push(entry.parent().unwrap().to_owned())
    }

    entries
}

fn clean_output(output: &str) -> String {
    #[derive(Default)]
    struct RawText(String);

    impl vte::Perform for RawText {
        fn print(&mut self, c: char) {
            self.0.push(c);
        }

        fn execute(&mut self, _: u8) {}

        fn hook(&mut self, _: &vte::Params, _: &[u8], _: bool, _: char) {}

        fn put(&mut self, b: u8) {
            self.0.push(b as char);
        }

        fn unhook(&mut self) {}

        fn osc_dispatch(&mut self, _: &[&[u8]], _: bool) {}

        fn csi_dispatch(&mut self, _: &vte::Params, _: &[u8], _: bool, _: char) {}

        fn esc_dispatch(&mut self, _: &[u8], _: bool, _: u8) {}
    }

    let mut raw = String::new();
    for line in output.lines() {
        let mut performer = RawText::default();
        let mut p = vte::Parser::new();
        for b in line.as_bytes() {
            p.advance(&mut performer, *b);
        }
        raw.push_str(&performer.0);
        raw.push('\n');
    }

    // Remove absolute paths from snapshot tests
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let manifest_dir: PathBuf = PathBuf::from(manifest_dir);
    let parent = manifest_dir.parent().unwrap();
    let result = raw.replace(&format!("{}/", parent.display()), "");

    // Remove compilation time
    let r = Regex::new("(Finished release \\[.*?\\] target\\(s\\) \\[.*?\\] in )(.*?s)").unwrap();
    let result = r.replace(&result, "$1???");

    result.to_string()
}
