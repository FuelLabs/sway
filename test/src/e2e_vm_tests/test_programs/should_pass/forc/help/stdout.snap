---
source: test/tests/tests.rs
---
> forc addr2line -h
exit status: 0
output:
Show location and context of an opcode address in its source file

Usage: forc addr2line [OPTIONS] --sourcemap-path <SOURCEMAP_PATH> --opcode-index <OPCODE_INDEX>

Options:
  -S, --search-dir <SEARCH_DIR>          Where to search for the project root [default: .]
  -g, --sourcemap-path <SOURCEMAP_PATH>  Source file mapping in JSON format
  -c, --context <CONTEXT>                How many lines of context to show [default: 2]
  -i, --opcode-index <OPCODE_INDEX>      Opcode index
  -v, --verbose...                       Use verbose output
  -s, --silent                           Silence all output
  -L, --log-level <LOG_LEVEL>            Set the log level
  -h, --help                             Print help

> forc build -h
exit status: 0
output:
Compile the current or target project

Usage: forc build [OPTIONS]

Options:
  -p, --path <PATH>
          Path to the project
      --offline
          Offline mode
  -t, --terse
          Terse mode
      --output-directory <OUTPUT_DIRECTORY>
          The directory in which Forc output artifacts are placed
      --locked
          Requires that the Forc.lock file is up-to-date
      --ipfs-node <IPFS_NODE>
          The IPFS node to use for fetching IPFS sources
      --ast
          Print the generated Sway AST (Abstract Syntax Tree)
      --dca-graph <DCA_GRAPH>
          Print the computed Sway DCA (Dead Code Analysis) graph
      --dca-graph-url-format <DCA_GRAPH_URL_FORMAT>
          URL format to be used in the generated DCA graph .dot file.
      --asm <ASM>...
          Print the generated ASM (assembler). [possible values: virtual, allocated, abstract, final, all]
      --bytecode
          Print the bytecode
      --ir <IR>...
          Print the generated Sway IR (Intermediate Representation). [possible values: initial, final, all, modified, inline, simplify-cfg, sroa, dce, fn-dce, fn-dedup-release, fn-dedup-debug, mem2reg, memcpyopt, const-folding, arg-demotion, const-demotion, ret-demotion, misc-demotion]
      --time-phases
          Output the time elapsed over each part of the compilation process
      --reverse-order
          Output build errors and warnings in reverse order
      --metrics-outfile <METRICS_OUTFILE>
          Output compilation metrics into the specified file
  -v, --verbose...
          Use verbose output
      --json-abi
          Minify JSON ABI files
  -s, --silent
          Silence all output
      --json-storage-slots
          Minify JSON storage slot files
  -L, --log-level <LOG_LEVEL>
          Set the log level
  -o, --output-bin <BIN_FILE>
          Create a binary file at the provided path representing the final bytecode
  -g, --output-debug <DEBUG_FILE>
          Create a file at the provided path containing debug information
      --build-profile <BUILD_PROFILE>
          The name of the build profile to use [default: debug]
      --release
          Use the release build profile
      --error-on-warnings
          Treat warnings as errors
      --build-target <BUILD_TARGET>
          Build target to use for code generation [default: fuel] [possible values: fuel, evm, midenvm]
      --tests
          Also build all tests within the project
      --experimental <EXPERIMENTAL>
          Comma separated list of all experimental features that will be enabled [possible values: new_encoding, storage_domains]
      --no-experimental <NO_EXPERIMENTAL>
          Comma separated list of all experimental features that will be disabled [possible values: new_encoding, storage_domains]
  -h, --help
          Print help (see more with '--help')
  -V, --version
          Print version

EXAMPLES:
    # Compile the current projectx
    forc build

    # Compile the current project from a different path
    forc build --path <PATH>

    # Compile the current project without updating dependencies
    forc build --path <PATH> --locked

> forc check -h
exit status: 0
output:
Check the current or target project and all of its dependencies for errors

Usage: forc check [OPTIONS] [BUILD_TARGET]

Arguments:
  [BUILD_TARGET]  Build target to use for code generation [default: fuel] [possible values: fuel, evm, midenvm]

Options:
  -p, --path <PATH>
          Path to the project, if not specified, current working directory will be used
      --offline
          Offline mode, prevents Forc from using the network when managing dependencies. Meaning it will only try to use previously downloaded dependencies
      --locked
          Requires that the Forc.lock file is up-to-date. If the lock file is missing, or it needs to be updated, Forc will exit with an error
  -t, --terse
          Terse mode. Limited warning and error output
      --disable-tests
          Disable checking unit tests
      --ipfs-node <IPFS_NODE>
          The IPFS Node to use for fetching IPFS sources
      --experimental <EXPERIMENTAL>
          Comma separated list of all experimental features that will be enabled [possible values: new_encoding, storage_domains]
      --no-experimental <NO_EXPERIMENTAL>
          Comma separated list of all experimental features that will be disabled [possible values: new_encoding, storage_domains]
  -v, --verbose...
          Use verbose output
  -s, --silent
          Silence all output
  -L, --log-level <LOG_LEVEL>
          Set the log level
  -h, --help
          Print help (see more with '--help')
  -V, --version
          Print version

EXAMPLES:
    # Check the current project
    forc check

    # Check the current project with a different path
    forc check --path <PATH>

    # Check the current project without updating dependencies
    forc check --locked

> forc clean -h
exit status: 0
output:
Removes the default forc compiler output artifact directory, i.e. `<project-name>/out`

Usage: forc clean [OPTIONS]

Options:
  -p, --path <PATH>            Path to the project, if not specified, current working directory will be used
  -v, --verbose...             Use verbose output
  -s, --silent                 Silence all output
  -L, --log-level <LOG_LEVEL>  Set the log level
  -h, --help                   Print help
  -V, --version                Print version

EXAMPLES:
    # Clean project
    forc clean

    # Clean project with a custom path
    forc clean --path <PATH>

> forc completions -h
exit status: 0
output:
Generate tab-completion scripts for your shell

Usage: forc completions [OPTIONS] --target <TARGET>

Options:
  -T, --target <TARGET>        Specify shell to enable tab-completion for [possible values: bash, elvish, fish, power-shell, zsh, fig]
  -v, --verbose...             Use verbose output
  -s, --silent                 Silence all output
  -L, --log-level <LOG_LEVEL>  Set the log level
  -h, --help                   Print help (see more with '--help')

> forc new -h
exit status: 0
output:
Create a new Forc project at `<path>`

Usage: forc new [OPTIONS] <PATH>

Arguments:
  <PATH>  The path at which the project directory will be created

Options:
      --contract               The default program type. Excluding all flags or adding this flag creates a basic contract program
      --script                 Adding this flag creates an empty script program
      --predicate              Adding this flag creates an empty predicate program
      --library                Adding this flag creates an empty library program
      --workspace              Adding this flag creates an empty workspace
      --name <NAME>            Set the package name. Defaults to the directory name
  -v, --verbose...             Use verbose output
  -s, --silent                 Silence all output
  -L, --log-level <LOG_LEVEL>  Set the log level
  -h, --help                   Print help
  -V, --version                Print version

EXAMPLES:
    # Create a new project
    forc new --contract --name my_project <PATH>

    # Create a new workspace
    forc new --workspace --name my_workspace <PATH>

    # Create a new Forc project with a predicate
    forc new <PATH> --predicate

    # Create a new Forc library project
    forc new <PATH> --library

> forc init -h
exit status: 0
output:
Create a new Forc project in an existing directory

Usage: forc init [OPTIONS]

Options:
      --path <PATH>            The directory in which the forc project will be initialized
      --contract               The default program type, excluding all flags or adding this flag creates a basic contract program
      --script                 Create a package with a script target (src/main.sw)
      --predicate              Create a package with a predicate target (src/predicate.rs)
      --library                Create a package with a library target (src/lib.sw)
      --workspace              Adding this flag creates an empty workspace
      --name <NAME>            Set the package name. Defaults to the directory name
  -v, --verbose...             Use verbose output
  -s, --silent                 Silence all output
  -L, --log-level <LOG_LEVEL>  Set the log level
  -h, --help                   Print help
  -V, --version                Print version

EXAMPLES:
    # Initialize a new Forc project
    forc init --path <PATH>

    # Initialize a new Forc project as workspace
    forc init --path <PATH> --workspace

    # Initialize a new Forc project with a predicate
    forc init --path <PATH> --predicate

    # Initialize a new Forc library project
    forc init --path <PATH> --library

> forc parse-bytecode -h
exit status: 0
output:
Parse bytecode file into a debug format

Usage: forc parse-bytecode [OPTIONS] <FILE_PATH>

Arguments:
  <FILE_PATH>  

Options:
  -v, --verbose...             Use verbose output
  -s, --silent                 Silence all output
  -L, --log-level <LOG_LEVEL>  Set the log level
  -h, --help                   Print help
  -V, --version                Print version

EXAMPLES:
    # Parse bytecode
    forc parse-bytecode <PATH>

> forc test -h
exit status: 0
output:
Run the Sway unit tests for the current project

Usage: forc test [OPTIONS] [FILTER]

Arguments:
  [FILTER]  When specified, only tests containing the given string will be executed

Options:
  -p, --path <PATH>
          Path to the project
      --offline
          Offline mode
  -t, --terse
          Terse mode
      --output-directory <OUTPUT_DIRECTORY>
          The directory in which Forc output artifacts are placed
      --locked
          Requires that the Forc.lock file is up-to-date
      --ipfs-node <IPFS_NODE>
          The IPFS node to use for fetching IPFS sources
      --ast
          Print the generated Sway AST (Abstract Syntax Tree)
      --dca-graph <DCA_GRAPH>
          Print the computed Sway DCA (Dead Code Analysis) graph
      --dca-graph-url-format <DCA_GRAPH_URL_FORMAT>
          URL format to be used in the generated DCA graph .dot file.
      --asm <ASM>...
          Print the generated ASM (assembler). [possible values: virtual, allocated, abstract, final, all]
      --bytecode
          Print the bytecode
      --ir <IR>...
          Print the generated Sway IR (Intermediate Representation). [possible values: initial, final, all, modified, inline, simplify-cfg, sroa, dce, fn-dce, fn-dedup-release, fn-dedup-debug, mem2reg, memcpyopt, const-folding, arg-demotion, const-demotion, ret-demotion, misc-demotion]
      --time-phases
          Output the time elapsed over each part of the compilation process
      --reverse-order
          Output build errors and warnings in reverse order
      --metrics-outfile <METRICS_OUTFILE>
          Output compilation metrics into the specified file
  -v, --verbose...
          Use verbose output
      --json-abi
          Minify JSON ABI files
  -s, --silent
          Silence all output
      --json-storage-slots
          Minify JSON storage slot files
  -L, --log-level <LOG_LEVEL>
          Set the log level
  -o, --output-bin <BIN_FILE>
          Create a binary file at the provided path representing the final bytecode
  -g, --output-debug <DEBUG_FILE>
          Create a file at the provided path containing debug information
      --build-profile <BUILD_PROFILE>
          The name of the build profile to use [default: debug]
      --release
          Use the release build profile
      --error-on-warnings
          Treat warnings as errors
      --build-target <BUILD_TARGET>
          Build target to use for code generation [default: fuel] [possible values: fuel, evm, midenvm]
      --pretty
          Pretty-print the logs emitted from tests
  -l, --logs
          Print `Log` and `LogData` receipts for tests
      --raw-logs
          Print the raw logs for tests
      --filter-exact
          When specified, only the test exactly matching the given string will be executed
      --test-threads <TEST_THREADS>
          Number of threads to utilize when running the tests. By default, this is the number of threads available in your system
      --experimental <EXPERIMENTAL>
          Comma separated list of all experimental features that will be enabled [possible values: new_encoding, storage_domains]
      --no-experimental <NO_EXPERIMENTAL>
          Comma separated list of all experimental features that will be disabled [possible values: new_encoding, storage_domains]
  -h, --help
          Print help (see more with '--help')
  -V, --version
          Print version

EXAMPLES:
    # Run test
    forc test

    # Run test with a filter
    forc test $filter

    # Run test without any output
    forc test --silent

    # Run test without creating or update the lock file
    forc test --locked

> forc update -h
exit status: 0
output:
Update dependencies in the Forc dependencies directory

Usage: forc update [OPTIONS]

Options:
  -p, --path <PATH>            Path to the project, if not specified, current working directory will be used
  -d <TARGET_DEPENDENCY>       Dependency to be updated. If not set, all dependencies will be updated
  -c, --check                  Checks if the dependencies have newer versions. Won't actually perform the update, will output which ones are up-to-date and outdated
      --ipfs-node <IPFS_NODE>  The IPFS Node to use for fetching IPFS sources
  -v, --verbose...             Use verbose output
  -s, --silent                 Silence all output
  -L, --log-level <LOG_LEVEL>  Set the log level
  -h, --help                   Print help (see more with '--help')
  -V, --version                Print version

EXAMPLES:
    # Update dependencies
    forc update

    # Update a specific dependency
    forc update -d std

    # Check if dependencies have newer versions
    forc update --check

> forc plugins -h
exit status: 0
output:
List all forc plugins

Usage: forc plugins [OPTIONS]

Options:
  -p, --paths                  Prints the absolute path to each discovered plugin
  -d, --describe               Prints the long description associated with each listed plugin
  -v, --verbose...             Use verbose output
  -s, --silent                 Silence all output
  -L, --log-level <LOG_LEVEL>  Set the log level
  -h, --help                   Print help (see more with '--help')
  -V, --version                Print version

EXAMPLES:
    # List all plugins
    forc plugins

    # List all plugins with their paths
    forc plugins --paths

    # List all plugins with their descriptions
    forc plugins --describe

    # List all plugins with their paths and descriptions
    forc plugins --paths --describe

> forc template -h
exit status: 0
output:
Create a new Forc project from a git template

Usage: forc template [OPTIONS] <PROJECT_NAME>

Arguments:
  <PROJECT_NAME>  The name of the project that will be created

Options:
  -u, --url <URL>                      The template url, should be a git repo [default: https://github.com/fuellabs/sway]
  -t, --template-name <TEMPLATE_NAME>  The name of the template that needs to be fetched and used from git repo provided
  -v, --verbose...                     Use verbose output
  -s, --silent                         Silence all output
  -L, --log-level <LOG_LEVEL>          Set the log level
  -h, --help                           Print help
  -V, --version                        Print version

EXAMPLES:
    # Create a new Forc project from an option template
    forc template new-path --template-name option

> forc contract-id -h
exit status: 0
output:
Determine contract-id for a contract. For workspaces outputs all contract ids in the workspace

Usage: forc contract-id [OPTIONS]

Options:
  -p, --path <PATH>
          Path to the project
      --offline
          Offline mode
  -t, --terse
          Terse mode
      --output-directory <OUTPUT_DIRECTORY>
          The directory in which Forc output artifacts are placed
      --locked
          Requires that the Forc.lock file is up-to-date
      --ipfs-node <IPFS_NODE>
          The IPFS node to use for fetching IPFS sources
      --json-abi
          Minify JSON ABI files
      --json-storage-slots
          Minify JSON storage slot files
      --ast
          Print the generated Sway AST (Abstract Syntax Tree)
      --dca-graph <DCA_GRAPH>
          Print the computed Sway DCA (Dead Code Analysis) graph
      --dca-graph-url-format <DCA_GRAPH_URL_FORMAT>
          URL format to be used in the generated DCA graph .dot file.
      --asm <ASM>...
          Print the generated ASM (assembler). [possible values: virtual, allocated, abstract, final, all]
      --bytecode
          Print the bytecode
      --ir <IR>...
          Print the generated Sway IR (Intermediate Representation). [possible values: initial, final, all, modified, inline, simplify-cfg, sroa, dce, fn-dce, fn-dedup-release, fn-dedup-debug, mem2reg, memcpyopt, const-folding, arg-demotion, const-demotion, ret-demotion, misc-demotion]
      --time-phases
          Output the time elapsed over each part of the compilation process
  -v, --verbose...
          Use verbose output
      --reverse-order
          Output build errors and warnings in reverse order
  -s, --silent
          Silence all output
  -L, --log-level <LOG_LEVEL>
          Set the log level
      --metrics-outfile <METRICS_OUTFILE>
          Output compilation metrics into the specified file
  -o, --output-bin <BIN_FILE>
          Create a binary file at the provided path representing the final bytecode
  -g, --output-debug <DEBUG_FILE>
          Create a file at the provided path containing debug information
      --build-profile <BUILD_PROFILE>
          The name of the build profile to use [default: debug]
      --release
          Use the release build profile
      --error-on-warnings
          Treat warnings as errors
      --salt <SALT>
          Added salt used to derive the contract ID
      --experimental <EXPERIMENTAL>
          Comma separated list of all experimental features that will be enabled [possible values: new_encoding, storage_domains]
      --no-experimental <NO_EXPERIMENTAL>
          Comma separated list of all experimental features that will be disabled [possible values: new_encoding, storage_domains]
  -h, --help
          Print help (see more with '--help')
  -V, --version
          Print version

EXAMPLES:
    # Get contract id
    forc contract-id

    # Get contract id from a different path
    forc contract-id --path <PATH>

> forc predicate-root -h
exit status: 0
output:
Determine predicate-root for a predicate. For workspaces outputs all predicate roots in the workspace

Usage: forc predicate-root [OPTIONS]

Options:
  -p, --path <PATH>
          Path to the project
      --offline
          Offline mode
  -t, --terse
          Terse mode
      --output-directory <OUTPUT_DIRECTORY>
          The directory in which Forc output artifacts are placed
      --locked
          Requires that the Forc.lock file is up-to-date
      --ipfs-node <IPFS_NODE>
          The IPFS node to use for fetching IPFS sources
      --json-abi
          Minify JSON ABI files
      --json-storage-slots
          Minify JSON storage slot files
      --ast
          Print the generated Sway AST (Abstract Syntax Tree)
      --dca-graph <DCA_GRAPH>
          Print the computed Sway DCA (Dead Code Analysis) graph
      --dca-graph-url-format <DCA_GRAPH_URL_FORMAT>
          URL format to be used in the generated DCA graph .dot file.
      --asm <ASM>...
          Print the generated ASM (assembler). [possible values: virtual, allocated, abstract, final, all]
      --bytecode
          Print the bytecode
      --ir <IR>...
          Print the generated Sway IR (Intermediate Representation). [possible values: initial, final, all, modified, inline, simplify-cfg, sroa, dce, fn-dce, fn-dedup-release, fn-dedup-debug, mem2reg, memcpyopt, const-folding, arg-demotion, const-demotion, ret-demotion, misc-demotion]
      --time-phases
          Output the time elapsed over each part of the compilation process
  -v, --verbose...
          Use verbose output
      --reverse-order
          Output build errors and warnings in reverse order
  -s, --silent
          Silence all output
  -L, --log-level <LOG_LEVEL>
          Set the log level
      --metrics-outfile <METRICS_OUTFILE>
          Output compilation metrics into the specified file
  -o, --output-bin <BIN_FILE>
          Create a binary file at the provided path representing the final bytecode
  -g, --output-debug <DEBUG_FILE>
          Create a file at the provided path containing debug information
      --build-profile <BUILD_PROFILE>
          The name of the build profile to use [default: debug]
      --release
          Use the release build profile
      --error-on-warnings
          Treat warnings as errors
      --experimental <EXPERIMENTAL>
          Comma separated list of all experimental features that will be enabled [possible values: new_encoding, storage_domains]
      --no-experimental <NO_EXPERIMENTAL>
          Comma separated list of all experimental features that will be disabled [possible values: new_encoding, storage_domains]
  -h, --help
          Print help (see more with '--help')
  -V, --version
          Print version

EXAMPLES:
    # Get predicate root
    forc predicate-root
